### PR TITLE
Fix Bug: raise exception "key, nr_ranks should be set at the same time" when calling `dist.all_reduce_sum(loss, "train_loss")`

### PR DESCRIPTION
**Pull request by wkcn**
在MegEngine文档中，`dist.all_reduce_sum`等函数允许nr_ranks为None
https://megengine.org.cn/doc/latest/_modules/megengine/distributed/functional.html#all_reduce_sum

然而当前版本在调用`dist.all_reduce_sum(loss, "train_loss")`时，会抛出异常："key, nr_ranks should be set at the same time"

本PR删除掉所有检测参数`nr_ranks`的语句

### CODE REVIEW

**Comment by Randolph87xxx** [comment]
你可以不填写key，dist.all_reduce_sum(loss)

对于多机算子，如果不填参数，默认按照init_process_group的配置填入key，nr_ranks，is_root等参数，如果要自己填写参数，需要把这些参数全部填写完整

**Comment by wkcn** [comment]
@Randolph87xxx 明白了，谢谢！
原因是我看到这个项目里用到了只填key不填nr_ranks的写法：https://github.com/megvii-model/WeightNet/blob/master/train.py#L156

**Comment by Randolph87xxx** [comment]
感谢你提出的意见，我们之后会对这个进行修改

**Comment by Randolph87xxx** [comment]
@wkcn 从用户的角度来看，你觉得这个注释怎么写比较方便理解，你可以在这个PR里进行修改

**Comment by ChaiMind** [comment]
> @wkcn 从用户的角度来看，你觉得这个注释怎么写比较方便理解，你可以在这个PR里进行修改

我认为值得讨论进一步讨论，针对该情况如何处理比较好，同样以该部分源码为例子：

```python
def all_reduce_sum(
    tensor: Tensor,
    key: Optional[str] = None,
    nr_ranks: Optional[int] = None,
    local_grad: Optional[bool] = False,
) -> Tensor:
    """Create all_reduce_sum operator for collective communication

    :param tensor: input tensor
    :param key: unique identifier for collective communication
    :param nr_ranks: number of ranks, use util.get_world_size() as default
    :param local_grad: whether use local grad
    """
    assert _group_check(key, nr_ranks), "key, nr_ranks should be set at the same time"
    return _collective_comm(
        tensor, key, Param.Mode.ALL_REDUCE_SUM, nr_ranks, local_grad=local_grad
    )
```

```python
@wrap_io_tensor
def _collective_comm(*args, **kargs):
    return collective_comm_symvar(*args, **kargs)
```

在 `help.py` 中有：

```python
def collective_comm_symvar(
    inp: Union[mgb.SymbolVar, mgb.CompGraph],
    key: Optional[str] = None,
    op: CollParam.Mode = None,
    nr_ranks: Optional[int] = None,
    is_root: Optional[bool] = None,
    rank: Optional[int] = None,
    local_grad: Optional[bool] = False,
    dtype: Optional[type] = None,
    device: Optional[mgb.CompNode] = None,
    comp_graph: Optional[mgb.CompGraph] = None,
) -> mgb.SymbolVar:
    """Helper function for creating collective_comm operators

    :param inp: tensor or comp_graph
    :param key: unique identifier for collective communication
    :param op: mode of collective communication
    :param nr_ranks: number of ranks, use util.get_world_size() as default
    :param is_root: whether this node is root node
    :param rank: rank of this node
    :param local_grad: whether use local grad
    :param dtype: output data type, use dtype of inp as default
    :param device: output comp node, use comp node of inp as default
    :param comp_graph: output comp graph, use comp graph of inp as default
    """
    return mgb.opr.collective_comm(
        inp,
        key=key if key is not None else ("collective_comm_" + str(get_group_id())),
        nr_devices=nr_ranks if nr_ranks is not None else get_world_size(),
        is_root=is_root if is_root is not None else (get_rank() == 0),
        rank=rank if rank is not None else get_rank(),
        local_grad=local_grad,
        server_addr=get_master_ip(),
        port=get_master_port(),
        param=CollParam(mode=op),
        dtype=dtype,
        backend=get_backend(),
        comp_node=device,
        comp_graph=comp_graph,
    )
```

根据以上源码可看出 @wkcn 单独去掉对参数 `nr_ranks` 的  `assert` 行为的做法也是不妥的。

在最新一次的代码改动中添加了对 `local_grad` 参数的设置以及 `assert _group_check()` 方法：

```python
def _group_check(*args):
    """Return True when arguments are all None or all not None
    """
    l = [val is None for val in args]
    return len(set(l)) <= 1
```

可以理解成，这个检查需要判断有没有对除了 `inp` 以外的参数进行设置。如果需要修改其它参数的值，就需要全部给出明确的实参。如果修改文档提醒用户除了 Input 以外的其它参数需要进行同时的自定义或者不定义，但可能会丧失部分对旧代码的兼容性，比如 WeightNet 的官方源码部分：

```python
https://github.com/megvii-model/WeightNet/blob/860c64bca8f75e54378f3885181eaaf5e66c9d46/train.py#L155-158
if dist.is_distributed():  # all_reduce_mean
            loss = dist.all_reduce_sum(loss, "train_loss") / dist.get_world_size()
            acc1 = dist.all_reduce_sum(acc1, "train_acc1") / dist.get_world_size()
            acc5 = dist.all_reduce_sum(acc5, "train_acc5") / dist.get_world_size()
```

此情况下，只需要传入 `key` 参数，而其它参数仍然可以采用 `collective_comm_symvar` 的初始化策略。

当然这也是由于作者没有在提供源码时指出 MegEngine 的具体版本，导致复现的时候 API 不兼容。

**Comment by Randolph87xxx** [comment]
是的，我觉得文档没有写好，如果写成
`key, nr_ranks should be set at the same time OR keep them None for default config`
感觉会更有好一些，但是觉得要写清楚就会很啰嗦
`key, nr_ranks should be set at the same time OR keep them None for default config, you can check it on megengine/distributed/help.py`

**Comment by wkcn** [comment]
谢谢你们详细的解释！
请问为什么需要在函数中提供nr_ranks参数的设置呢？什么情况下，nr_ranks会不等于world_size呢？

是不是有这种使用情况：
8个训练进程，all_reduce_sum的nr_ranks设为4，此时只要有4个进程调用了all_reduce_sum，就可以直接求和，把结果给这4个进程。

**Comment by ChaiMind** [comment]
> 谢谢你们详细的解释！
> 请问为什么需要在函数中提供nr_ranks参数的设置呢？什么情况下，nr_ranks会不等于world_size呢？
> 
> 是不是有这种使用情况：
> 8个训练进程，all_reduce_sum的nr_ranks设为4，此时只要有4个进程调用了all_reduce_sum，就可以直接求和，把结果给这4个进程。

你考虑的可能是多卡的情况，开发者在设计时还要考虑多机的情况。

**Comment by wkcn** [comment]
我不更新这个PR了，谢谢你们～
